src/bindings.rs

```diff
@@ -161,40 +136,25 @@ impl Display for Shortcut {
     }
 }
 
-struct ShortcutVisitor;
-
-impl<'de> de::Visitor<'de> for ShortcutVisitor {
-    type Value = Shortcut;
-
-    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
-        formatter.write_str("a shortcut identifier")
-    }
+impl FromStr for Shortcut {
+    type Err = ShortcutFromStrError;
 
-    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
-    where
-        E: de::Error,
-    {
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
         let keys: Vec<&str> = value.split('-').collect::<Vec<&str>>();;
         if let Some((key_code_string, modifier_strings)) = keys.split_last() {
-            let modifiers_state = ModifiersState {
+            let modifiers = ModifiersState {
                 ctrl: modifier_strings.contains(&"C"),
                 alt: modifier_strings.contains(&"M"),
                 shift: modifier_strings.contains(&"S"),
                 logo: modifier_strings.contains(&"L"),
             };
-            let key_code_result = key_code_from_str(key_code_string);
-            match key_code_result {
-                Ok(key_code) => Ok(Shortcut {
-                    modifiers: modifiers_state,
-                    key_code: key_code,
-                }),
-                Err(error) => Err(error),
-            }
+            let key_code = key_code_string.parse()?;
```

Comment:
Wut, how does Rust know into what type it should parse the value? Does it implicitely pull the type from the `Shortcut` return a line below?

Reply:
Basically, yeah. Type inference works in both directions, like in Haskell etc. :)